### PR TITLE
add v,dAdt as postprocessing field.

### PR DESCRIPTION
**Pull request by 0nu**
Hi Guilherme,

we need v and dAdt to be passed through to the postprocessing function that we use for TMS coil optimizations.

A more generic way to build the `out_field` would also be nice, but for now the patch below would work for us.

Cheers,
Ole

### CODE REVIEW

File: simnibs/simulation/fem.py [hunk 1]
```diff
@@ -1410,8 +1410,8 @@ def tms_many_simulations(
         Regions of interest where the fields is to be saved.
         If set to None, will save the electric field in all tissues.
         Default: None
-    field: 'E' or 'J' (optional)
-        Which field to save (electric field E or current density J). Default: 'E'
+    field: 'E', 'J', 'vdAdt' (optional)
+        Which field to save (electric field E, current density J, or v and dAdt). Default: 'E'
```
Comment (0nu) [inline simnibs/simulation/fem.py:1414]: Doc string

File: simnibs/simulation/fem.py [hunk 2]
```diff
@@ -1421,8 +1421,8 @@ def tms_many_simulations(
     n_workers: int
         Number of workers to use
     '''
-    if field != 'E' and field != 'J':
-        raise ValueError("Field shoud be either 'E' or 'J'")
+    if field != 'E' and field != 'J' and field.lower() != 'vdadt':
+        raise ValueError("Field can be either 'E', 'J', or 'vdAdt'")
```
Comment (0nu) [inline simnibs/simulation/fem.py:1425]: adapt parameter check

File: simnibs/simulation/fem.py [hunk 3]
```diff
@@ -1467,6 +1467,8 @@ def tms_many_simulations(
                 out_field = E
             elif field == 'J':
                 out_field = calc_J(E, cond)
+            elif field.lower() == 'vdadt':
+                out_field = (v, dAdt)
```
Comment (0nu) [inline simnibs/simulation/fem.py:1471]: add v,dadt for single core

File: simnibs/simulation/fem.py [hunk 4]
```diff
@@ -1540,6 +1542,8 @@ def _run_tms_many_simulations(i, matsimnibs, didt, fn_hdf5, dataset):
         out_field = E
     elif tms_many_global_field == 'J':
         out_field = calc_J(E, tms_many_global_cond)
+    elif tms_many_global_field.lower() == 'vdadt':
+        out_field = (v,dAdt)
```
Comment (0nu) [inline simnibs/simulation/fem.py:1546]: add add v,dadt for multi core

**Comment by oulap** [comment]
Hi Ole,

Looks reasonable, but could you change the string for dA/dt to 'D' so that's consistent with the GUI.
Also now that you're anyway returning multiple out_fields maybe change the code to parse the whole string so we could for example define 'EvD' to get E, v and dA/dt.

Best,
Oula

File: simnibs/simulation/fem.py [hunk 5]
```diff
@@ -1463,12 +1462,22 @@ def tms_many_simulations(
...
             E -= dAdt
-            if field == 'E':
-                out_field = E
-            elif field == 'J':
-                out_field = calc_J(E, cond)
-            elif field.lower() == 'vdadt':
-                out_field = (v, dAdt)
+
+            # build output fields
+            out_field = []
```
Comment (0nu) [inline simnibs/simulation/fem.py:1467]: each requested field is added to out_field list...

File: simnibs/simulation/fem.py [hunk 6]
```diff
@@ -1463,12 +1462,22 @@ def tms_many_simulations(
...
+            if 'E' in field:
+                out_field.append(E)
+            if 'D' in field:
+                out_field.append(dAdt)
+            if 'J' in field:
+                out_field.append(calc_J(E, cond))
+            if 'v' in field:
+                out_field.append(v)
+            out_field = tuple(out_field)
```
Comment (0nu) [inline simnibs/simulation/fem.py:1476]: ...made into tuple for backward compatibility....

File: simnibs/simulation/fem.py [hunk 7]
```diff
@@ -1463,12 +1462,22 @@ def tms_many_simulations(
...
+
+            # if only one field to output, un-tuple
+            if len(out_field) == 1:
+                out_field = out_field[0]
```
Comment (0nu) [inline simnibs/simulation/fem.py:1480]: ...and if only one field (the 'standard' way), untupled, to keep compatibility with old post_pro functions

File: simnibs/simulation/fem.py [hunk 8]
```diff
@@ -1463,10 +1462,22 @@ def tms_many_simulations(
...
             if post_pro is not None:
                 out_field = post_pro(out_field)
             with h5py.File(fn_hdf5) as f:
```
Comment (0nu) [inline simnibs/simulation/fem.py:1483]: I guess calling tms_many_simulations() with the new feature of multiple fields (e.g. with field='ED') but without a post_pro function would crash here. To circumvent, one would probably have to change the structure of fn_hdf5 here and everywhere.

**Comment by oulap** [comment]
Thanks, I'll add a check for the post-processing so that the code throws an error if it's called with post-processing but the wrong fields.

**Pull request merged**